crates/noirc_evaluator/src/ssa/anchor.rs

```diff
@@ -0,0 +1,263 @@
+use std::collections::{HashMap, VecDeque};
+
+use super::{
+    context::SsaContext,
+    mem::ArrayId,
+    node::{NodeId, ObjectType, Opcode, Operation},
+};
+
+#[derive(Clone, Debug)]
+pub enum MemItem {
+    NonConst(NodeId),
+    Const(usize),     //represent a bunch of memory instructions having constant index
+    ConstLoad(usize), //represent a bunch of load instructions having constant index
+}
+
+#[derive(Debug)]
+pub enum CseAction {
+    ReplaceWith(NodeId),
+    Remove(NodeId),
+    Keep,
+}
+
+/// A list of instructions with the same Operation variant
+#[derive(Default, Clone)]
+pub struct Anchor {
+    map: HashMap<Opcode, HashMap<Operation, NodeId>>, //standard anchor
+    cast_map: HashMap<NodeId, HashMap<crate::node::ObjectType, NodeId>>, //cast anchor
+    mem_map: HashMap<ArrayId, Vec<VecDeque<(usize, NodeId)>>>, //Memory anchor: one Vec for each array where Vec[i] contains the list of load and store instructions having index i, and the mem_item position in which they appear
+    mem_list: HashMap<ArrayId, VecDeque<MemItem>>, // list of the memory instructions, per array, and grouped into MemItems
+}
+
+impl Anchor {
+    pub fn push_cast_front(&mut self, operation: &Operation, id: NodeId, res_type: ObjectType) {
+        match operation {
+            Operation::Cast(cast_id) => {
+                let mut by_type = HashMap::new();
+                by_type.insert(res_type, id);
+                self.cast_map.insert(*cast_id, by_type);
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    pub fn push_front(&mut self, operation: &Operation, id: NodeId) {
```

Comment:
Why separate this from `push_cast_front`?

Reply:
I could use one function but for Cast instructions I need the result type and I felt it would be awkward to require it for the other cases. I don't have strong preference though.

Reply:
This is fine, I didn't see the extra parameter originally since it was wrapped onto the next line on github